lib/reversetunnel/track/tracker.go

```diff
@@ -52,268 +44,307 @@ func (c *Config) CheckAndSetDefaults() error {
-func (t *Tracker) Start() {
-	t.mu.Lock()
-	defer t.mu.Unlock()
-	t.getOrCreate()
+func (t *Tracker) expireProxiesLocked() {
+	now := time.Now()
+	for k, v := range t.tracked {
+		if v.expiry.Before(now) {
+			delete(t.tracked, k)
+			t.cannotLease = false
+		}
+	}
 }
 
-// Stop stops tracking for specified proxy address.
-func (t *Tracker) Stop() {
-	t.mu.Lock()
-	defer t.mu.Unlock()
-	if t.sets == nil {
-		return
+// canLeaseLocked returns true if the current state of the Tracker is such that
+// the client should attempt to spawn new connections.
+func (t *Tracker) canLeaseLocked() bool {
+	// degenerate condition: we've just started the tracker or we haven't
+	// successfully connected to any server yet, or network conditions were such
+	// that all tracked proxies have expired - spawn one connection, or we'll
+	// get stuck like that forever
+	if t.inflight == 0 && (len(t.claimed) == 0 || len(t.tracked) == 0) {
+		return true
 	}
 
-	t.sets = nil
-	t.wp.Set(0)
-}
-
-// StopAll permanently deactivates this tracker and cleans
-// up all background goroutines.
-func (t *Tracker) StopAll() {
-	t.cancel()
-}
+	desiredGen := make(map[string]uint64, 8)
+	for _, v := range t.tracked {
+		if v.Generation > desiredGen[v.Group] {
+			desiredGen[v.Group] = v.Generation
+		}
+	}
 
-func (t *Tracker) tick() {
-	t.mu.Lock()
-	defer t.mu.Unlock()
-	if t.sets == nil {
-		return
+	desired := make(map[string]struct{}, len(t.tracked))
+	for k, v := range t.tracked {
+		if desiredGen[v.Group] == v.Generation {
+			desired[k] = struct{}{}
+		}
 	}
 
-	cutoff := time.Now().Add(-1 * t.ProxyExpiry)
-	if t.sets.expire(cutoff) > 0 {
-		count := len(t.sets.proxies)
-		if count < 1 {
-			count = 1
+	// do the set intersection like this because claimed is generally going to
+	// be smaller than desired/tracked
+	desiredClaimed := 0
+	for k := range t.claimed {
+		if _, ok := desired[k]; ok {
+			desiredClaimed++
 		}
-		t.wp.Set(uint64(count))
 	}
-}
 
-func (t *Tracker) getOrCreate() *proxySet {
-	if t.sets == nil {
-		t.sets = newProxySet(t.ClusterName)
-		t.wp.Set(1)
+	desiredCount := len(desired)
+	if t.connectionCount > 0 && desiredCount > t.connectionCount {
+		desiredCount = t.connectionCount
 	}
 
-	return t.sets
+	return desiredCount > desiredClaimed+t.inflight
 }
 
-// Claim attempts to claim a lease based on the given principals returning a
-// function to unclaim and if the claim was successful.
-func (t *Tracker) Claim(principals ...string) (unclaim func(), ok bool) {
-	if ok := t.claim(principals...); !ok {
-		return nil, false
+// TrackExpected starts/refreshes tracking for expected proxies.  Called by
+// agents when gossip messages are received.
+func (t *Tracker) TrackExpected(proxies ...Proxy) {
+	if len(proxies) == 0 {
+		return
+	}
+
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	expiry := time.Now().Add(t.proxyExpiry)
+	t.cannotLease = false
+	for _, p := range proxies {
+		p.expiry = expiry
+		t.tracked[p.Name] = p
```

Comment:
nit: consider adding a warning if a new proxy is observed that has a different generation value than a previous proxy with the same ID (when thinking through likely misuse/misconfiguration scenarios, reusing the same proxy IDs across multiple generations strikes me as one of the easier ways this system might get messed up by user error).

Reply:
Group IDs and generations are logged at debug level by the agent code handling discovery requests so while this feature is unstable and only used by cloud, I think we're going to be able to diagnose issues just from that: 

{"caller":"reversetunnel/agent.go:648","component":"proxy:agent","leaseID":24,"level":"debug","message":"Received discovery request: [0b540107-8212-4e4b-a726-6e94cae5b9c6(us-east-1@21), 1c7583eb-865c-4700-a6a2-44025ecc613c(us-west-2@20), 5b2ce3cc-63ea-463c-8e30-14b34cba62e9(us-west-2@20), d45bc8c3-6522-4b15-a7ca-fa9e6cddc5b7(us-east-1@21)]","target":"espadolini.trunk.cloud.teleportinfra.dev:443","timestamp":"2023-07-10T11:26:07+02:00","trace.fields":{"localCluster":"","targetCluster":"espadolini.trunk.cloud.teleportinfra.dev"}}

I'll leave a TODO for better logging of that condition if and when we make the feature public.